Added JavaScript-Classes.md

alwinsol opened the pull request:
Adding A new README on JavaScript Classes.  Originally written as a today-i-learned file.  The content proved to be more than a quick read so references added and updated.

alwinsol committed:
Added JavaScript-Classes.md

lucymonie commented on JavaScript/JavaScript-Classes.md:
```diff
@@ -0,0 +1,171 @@
+
+###Introduction
+
+JavaScript is a prototypal object-oriented language, meaning that objects can inherit directly from other objects.  This is a bit different to those who have learned to develop in other langauges like Java or C# where a subclass will inherit from a superclass, and objects are instances of classes. 
```
@alwinsol, you say 'objects can inherit directly from other objects'. Do you mean 'objects can inherit properties directly from other objects'? It would be good to be more specific here about what is inherited.
Also, in the language comparison, you could express this more concisely as:

> This is a bit different from  languages like Java or C# where a subclass inherits from a superclass, and objects are instances of classes.

alwinsol replied:
Amended

lucymonie commented on JavaScript/JavaScript-Classes.md:
```diff
@@ -0,0 +1,171 @@
+
+Each of these four is a different approach to object construction; they all accomplish the same thing in a slightly different way.
+
+Take for example, a street, which has many houses. Every house on the street requires a colour for the house, a front door with methods to open and close the door. A very simple example of where houses may be thought of as objects. 
```
This paragraph might be a little clearer as:

> Take for example, a street with many houses. Every house on the street has a colour, and a front door that must open and close. So houses may be thought of as objects with a colour property and an open-and-close-door method.

alwinsol replied:
Amended

lucymonie commented on JavaScript/JavaScript-Classes.md:
```diff
@@ -0,0 +1,171 @@
+
+Take for example, a street, which has many houses. Every house on the street requires a colour for the house, a front door with methods to open and close the door. A very simple example of where houses may be thought of as objects. 
+
+We can implement functional classes here, as every instance created by the function will have the attributes of house colour and the methods to open and close the door.  
```
And this paragraph would then be something more like:

> If we implement functional classes here, every instance created by the function will have the attributes of house colour and the method to open and close the door.

alwinsol replied:
Amended

lucymonie commented on JavaScript/JavaScript-Classes.md:
```diff
@@ -0,0 +1,171 @@
+
+We can implement functional classes here, as every instance created by the function will have the attributes of house colour and the methods to open and close the door.  
+
+From here we can potentially create hundreds, thousands or even more new house objects, which are similar.  So each functional class performs the same basic steps to generate the new houses.
```
Suggest:

> From here we can potentially create unlimited new house objects. The functional class performs the same basic steps to generate the new houses:

alwinsol replied:
Amended

lucymonie commented on JavaScript/JavaScript-Classes.md:
```diff
@@ -0,0 +1,171 @@
+Each of the four styles has its pros and cons.
+
+####Functional
+Firstly **functional style** is **fast, very simple and clear in object creation**.  However all methods will be duplicated for every single object created, so a new function is stored in memory every time the object is generated.
```
Suggest changing:

> However all methods will be duplicated for every single object created, and a new function is stored in memory every time the object is generated. If you are working with large numbers of objects, this may not be the best choice.

alwinsol replied:
Changed

lucymonie commented on JavaScript/JavaScript-Classes.md:
```diff
@@ -0,0 +1,171 @@
+};
+
+houseMethods.closeDoor = function(){
+	this.door = 'closed'; //Add method to delegate fallback object
```
Should this be 'close' rather than 'closed'?

```
this.door = 'close';
```

alwinsol replied:
Amended - thanks for checking

lucymonie commented on JavaScript/JavaScript-Classes.md:
`````diff
@@ -0,0 +1,171 @@
+var house = House('red'); //instantiation pattern
+
+````
+This is not as efficient as delegating through fallback.
`````
I am not quite clear here what you mean by 'delegating'. Does this mean retrieving properties from the prototype?

alwinsol replied:
amended

lucymonie commented on JavaScript/JavaScript-Classes.md:
```diff
@@ -0,0 +1,171 @@
+This is not as efficient as delegating through fallback.
+
+####Prototypal
+Fallbacks are a back-up plan for objects, they are the cornerstone of **prototypal** style.  If a method or property is called on an object, where it does not exist for that object, JavaScript will check if it's defined on its fallback object. (See the README on [Inheritance and JavaScript](https://github.com/codingforeveryone/READMEs/blob/master/JavaScript/inheritance-and-javascript.md)).  Every ...
```
Suggest moving first line of this para to middle of paragraph like so:

> If a nonexistent object method or property is called, JavaScript will check if it's defined on its fallback object. (See the README on [Inheritance and JavaScript](https://github.com/codingforeveryone/READMEs/blob/master/JavaScript/inheritance-and-javascript.md)).  Fallbacks are like a back-up plan for objects, they are the cornerstone of **prototypal** style.  Every function has a property called prototype where the fallback methods are stored.

alwinsol replied:
Moved

lucymonie commented on JavaScript/JavaScript-Classes.md:
```diff
@@ -0,0 +1,171 @@
+};
+
+House.prototype.closeDoor = function(){
+	this.door = 'closed';
```
Again, should this be 

```
this.door = 'close';
```

alwinsol replied:
changed

lucymonie commented on JavaScript/JavaScript-Classes.md:
```diff
@@ -0,0 +1,171 @@
+Note that Pseudoclassical is the only style that makes use of the `new` keyword, the other three classes are used like a function call.
+
+####Summary
+None of the above are better or worse to use, they are just styles, and it is up to you as the programmer to decide which your program would benefit the most from.
```
This summary would be a great place to briefly restate the case for each style. Or just say 'For example if you are working with very large numbers of examples, you might choose function-shared rather than functional.'

alwinsol replied:
Added pros and cons.

alwinsol committed:
Updated JavaScript Classes Document

lucymonie commented:
This looks good to me. It's probably good if someone else proofreads it and then merges it.

Cleop commented:
I'm happy to proofread and merge! I'll give it a look now :-)

Cleop committed:
Proofreading changes

Great piece, I certainly learnt a few things! I've renamed the variable House as house as it's recommended to name variables beginning with a lower case letter (http://www.w3schools.com/js/js_conventions.asp). I've also rephrased a couple of bits slightly just to help them read a little clearer and changed some spelling/ grammatical errors. I've also introduced some syntax highlighting using markdowns.  Let me know if anything doesn't make sense!

Cleop committed:
2nd attempt!

Now with House objects as they correctly were before with a capital C! Looks ready to merge in my opinion.

msachi commented on JavaScript/JavaScript-Classes.md:
```diff
@@ -0,0 +1,193 @@
+    }; //explicitly define or borrow methods
+
+    obj.closeDoor = function(){
+        obj.door = 'close;
```
Missing quotation mark: obj.door = 'close';

alwinsol replied:
done

msachi commented on JavaScript/JavaScript-Classes.md:
```diff
@@ -0,0 +1,193 @@
+	};
+
+	houseMethods.closeDoor = function(){
+    this.door = 'closed'; //Add method to delegate fallback object
```
Use 'close' instead of 'closed' for consistency throughout the document

alwinsol replied:
done

msachi commented on JavaScript/JavaScript-Classes.md:
```diff
@@ -0,0 +1,193 @@
+	};
+
+	House.prototype.closeDoor = function(){
+    this.door = 'closed';
```
Use 'close' again

alwinsol replied:
done

msachi commented on JavaScript/JavaScript-Classes.md:
```diff
@@ -0,0 +1,193 @@
+
+* Often the most optimized pattern and very common
+
+* *this* can be difficult to understand
```
Maybe say 'keyword this' to avoid someone thinking that 'this' refers to the pseudoclassical style. (You have it in italics, but it would be even clearer like that.)

alwinsol replied:
done

msachi commented on JavaScript/JavaScript-Classes.md:
```diff
@@ -0,0 +1,193 @@
+
+[Instantiation patterns in JavaScript](http://callmenick.com/post/instantiation-patterns-in-javascript)
+
+JavaScript: The Good Parts - Douglas Crockford Chapter 5
```
Might be nice to add a link to the PDF document, e.g. http://bdcampbell.net/javascript/book/javascript_the_good_parts.pdf

alwinsol replied:
In case this is a 'non-legitimate' copy link is not included.

alwinsol committed:
Update JavaScript-Classes.md

Incorporated latest comments

lucymonie commented:
I just proof-read this, and it's looking fantastic. Well done @alwinsol!
